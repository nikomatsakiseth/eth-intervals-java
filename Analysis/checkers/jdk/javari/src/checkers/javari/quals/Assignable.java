package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;

import static java.lang.annotation.ElementType.*;

/**
 * Indicates that a field is assignable, even if it is inside a {@link ReadOnly}
 * instance.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({FIELD})
public @interface Assignable {

}
