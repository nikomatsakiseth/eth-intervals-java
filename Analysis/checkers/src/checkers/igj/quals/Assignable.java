package checkers.igj.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.*;

/**
 *
 * Indicates the annotated {@code Field} may be re-assigned regardless of the
 * immutability of the enclosing class or object instance.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target( { FIELD } )
public @interface Assignable {

}
