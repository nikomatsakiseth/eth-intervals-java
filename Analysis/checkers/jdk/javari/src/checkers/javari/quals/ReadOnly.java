package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;
import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import static java.lang.annotation.ElementType.*;

/**
 * Indicates that, for the variable on which this annotation appears,
 * the object to which this variable refers will not be modified via
 * this reference, except its fields explicitly marked as
 * {@link Mutable}.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see Mutable
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target({FIELD, LOCAL_VARIABLE, METHOD, PARAMETER, TYPE})
@TypeQualifier
@SubtypeOf({})
public @interface ReadOnly {

}
