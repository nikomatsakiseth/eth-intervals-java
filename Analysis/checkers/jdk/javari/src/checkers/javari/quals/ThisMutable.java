package checkers.javari.quals;

import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 * An annotation used to represent a place holder immutability type, that is
 * equivalent to the ThisMutable type on the Javari typesystem.
 *
 * However, it is an implementation detail; hence, the package-scope.
 */
@TypeQualifier
@Target({}) // A programmer cannot write @ThisMutable in a program
@SubtypeOf(ReadOnly.class)
public @interface ThisMutable {

}
