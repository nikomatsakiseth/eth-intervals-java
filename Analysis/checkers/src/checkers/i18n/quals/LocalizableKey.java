package checkers.i18n.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import checkers.quals.Unqualified;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface LocalizableKey { }
