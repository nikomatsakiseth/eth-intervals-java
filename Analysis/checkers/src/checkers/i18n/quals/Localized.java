package checkers.i18n.quals;

import static com.sun.source.tree.Tree.Kind.*;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.*;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@ImplicitFor( trees = {
        /* All integer literals */
        INT_LITERAL,
        LONG_LITERAL,
        FLOAT_LITERAL,
        DOUBLE_LITERAL,
        BOOLEAN_LITERAL,

        //CHAR_LITERAL,
        //STRING_LITERAL,
        //NULL_LITERAL
        }
)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface Localized { }
