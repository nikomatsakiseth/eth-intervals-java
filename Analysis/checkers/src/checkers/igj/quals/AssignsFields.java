package checkers.igj.quals;

import static java.lang.annotation.ElementType.*;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 *
 * Indicates that the annotated method could assigns (but not mutate) the fields
 * of {@code this} object.
 *
 * A method with an AssignsFields receiver may not use the receiver to
 * invoke other methods with mutable or immutable reciever.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
// TODO: Document this

@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target( { CONSTRUCTOR, METHOD })
@TypeQualifier // (for now)
@SubtypeOf( ReadOnly.class )
public @interface AssignsFields {

}
