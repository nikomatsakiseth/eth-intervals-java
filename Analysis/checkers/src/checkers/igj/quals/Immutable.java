package checkers.igj.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.quals.*;

import static java.lang.annotation.ElementType.*;

/**
 * Indicates that the annotated reference is an immutable reference to an
 * immutable object.
 *
 * An Immutable object cannot be modified. Its fields may be reassigned or
 * mutated only if they are explicitly marked as Mutable or Assignable.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target( { FIELD, LOCAL_VARIABLE, CONSTRUCTOR, METHOD, PARAMETER, TYPE })
@TypeQualifier
@SubtypeOf( ReadOnly.class )
public @interface Immutable {

}
