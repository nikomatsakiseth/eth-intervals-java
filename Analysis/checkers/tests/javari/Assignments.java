import checkers.javari.quals.*;
import java.awt.Point;

class Assignments {

    Point a = new Point(0, 1);
    @ReadOnly Point b = new Point(2, 3),
        c = new Point(4, 5);
    int i = 0;

    Assignments aMutable;
    @ReadOnly Assignments aReadOnly;

    String mString;
    @ReadOnly String roString;

    class JavariCell {
        JavariCell cell;
        int x;

        public void mutateInternal(JavariCell cell) {
            this.cell = cell;
        }

        public void requiresMutableParameter(JavariCell cell) @ReadOnly {}

        @PolyRead JavariCell getCell() @PolyRead {
            return cell;
        }
    }

    JavariCell mc;
    @ReadOnly JavariCell roc;

    @Mutable public String isMutable() @ReadOnly {
        return "isMutable";
    }

    public String isThisMutable() @ReadOnly {
        return "isThisMutable";
    }

    @ReadOnly public String isReadOnly() @ReadOnly {
        return "isReadOnly";
    }

    @PolyRead public String isPolyRead(@PolyRead Object c) {
        return "isPolyRead";
    }

    public void canDo() {
        i = b.x;
        c = b;
        b = a;
        b = (@ReadOnly Point) a;
        a = (Point) b; // should emit a warning here
        a.x = 0;
        i = a.x;
        a.x = b.y;   // can assign PRIMITIVE field
        mc.cell = mc;
        roc = roc.cell;
        roc = null;
        mc.mutateInternal(mc);
        mc.requiresMutableParameter(mc);
        roc.requiresMutableParameter(mc);
        mc = mc.getCell();
        roc = roc.getCell();

        roString = isPolyRead(roString); // polyread resolved as readonly
        mString = isPolyRead(mString);   // polyread resolved as mutable;
        roString = isPolyRead(mString);  // polyread resolved as mutable;

        mString = isMutable();
        mString = isThisMutable();

        roString = isMutable();
        roString = isThisMutable();
        roString = isReadOnly();


        // mutable can be assigned to mutable
        mString = aMutable.isMutable();
        mString = aMutable.isThisMutable();
        mString = aReadOnly.isMutable();

        // anything can be passed to a readonly
        roString = aMutable.isMutable();
        roString = aMutable.isThisMutable();
        roString = aMutable.isReadOnly();
        roString = aReadOnly.isMutable();
        roString = aReadOnly.isThisMutable();
        roString = aReadOnly.isReadOnly();
    }

    public void doNothing() @ReadOnly {
        JavariCell localCell = null;      // mutable
        localCell.x = 3;
        localCell.cell = localCell;            // is mutable, no error
    }

    public int getInt() @ReadOnly { return 0; }
    public void assignmentWithCast() @ReadOnly {
        char c = (char) getInt();
    }
}
