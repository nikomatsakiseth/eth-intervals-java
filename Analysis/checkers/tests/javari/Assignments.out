:56: warning: (cast.unsafe)